Peer deployment guide (#1803)
hyperledger/fabric: docs/source/deploypeer/peerdeploy.md

**Tip:** Add the location of the peer binary to your `PATH` environment variable so that it can be picked up without fully qualifying the path to the binary executable, for example:

```
export PATH=<path to download location>/bin:$PATH
```

After you have mastered deploying and running a peer by using the peer binary and core.yaml configuration file, it is likely that you will want to use a peer container in a Kubernetes or Docker deployment. The Hyperledger Fabric project publishes a [peer image](https://hub.docker.com/r/hyperledger/fabric-peer) that can be used for development and test, and various vendors provide supported peer images. For now though, the purpose of this topic is to teach you how to properly use the peer binary so you can take that knowledge and apply it to the production environment of your choice.

## Prerequisites

Before you can launch a peer node in a production network, you need to configure it. Complete the following steps:

> 

Reviewer: The flow is awkward here.
Prereq steps 1,2,3 are all about certs (While Step 2 is only a note, and lists the substeps of Step 3, strange). Then step 3 has many sub-steps and sub-sub-steps.
Step 4 is about storage.
Step 5 and 6 are about config.

Instead of numbered steps 1 through 6, I'd suggest 3 sections:
Certificates (Include the content from current steps 1,2,3)
Storage (Include the content from current step 4)
Configuration (Include the content from current steps 5,6).

Each section can then have a direct link, and less nesting of sub bullets.

                      └── tls
    ```
2. While **cryptogen** is a convenient utility that can be used to generate certificates for a test network, it should **never** be used on a production network. The core requirement for certificates for Fabric nodes is that they are Elliptic Curve (EC) certificates. You can use any tool you prefer to issue these certificates (for example, OpenSSL). However, the Fabric CA streamlines the process because it generates the Membership Service Providers (MSPs) for you. Regardless of which certificate provider you choose, you should have already used it to generate the peer enrollment certificate, TLS certificate, private keys, and the MSPs that Fabric must consume. Refer to the [CA deployment](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/cadeploy.html) and [Registering and enrolling identities with a CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html) topics for instructions on how to generate these certificates. You need to generate the following sets of certificates:
    - **Peer TLS CA certificates**
    - **Peer organization certificates (organization MSP)**

>     - **Peer enrollment certificates (peer local MSP)**  

Reviewer: Make the bullet name consistent with below, I'd recommend "Peer local MSP (enrollment certificate and private key)".

      - **TLS certificates**:
        - Copy the TLS CA TLS certificate `tls-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/tls-cert.pem`. The path and name of the certificate corresponds to `peer.tls.rootcert.file` parameter in `core.yaml`.
        - After you have generated the peer TLS certificate, the certificate will have been generated in the `signcerts` directory, and the private key will have been generated in the `keystore` directory. Rename the generated private key in the `keystore` folder to `peer0-key.pem` so that it can more easily be recognized later on as being a private key.
        - Copy the peer TLS certificate and private key to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls`. The path and name of the certificate and private key files correspond to the `peer.tls.cert.file` and `peer.tls.key.file` parameters in the `core.yaml`.

>       - **Peer organization MSP**:

Reviewer: I'm thinking this whole bullet and sub-bullets can be removed now.
We're not configuring the org MSP here, just the peer's local MSP.  The org MSP would be configured when creating the channel.
Or am I missing the intent?

Author: I agree. I think Pam added this because it was part of her process for deploying the peer (since it is a prerequisite). But it's covered in the CA docs so I think we can probably scrap this little section.

What do you think @pamandrejko?

Reviewer 2: I checked the [Fabric CA deployment guide](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#folder-structure-for-your-org-and-node-admin-identities) and this specific information is not in there. It just says to create the `tls` folder in the folder structure. I think the intent was to fill in the TLS  certs later when you create the node. While you use the CA to register and enroll the peer identity, when you deploy the peer, you have to build out this folder structure, and because this is a tutorial I am providing the specific instructions how to do this. 

So I suppose we have a choice then, include it in here, or add more details to the CA deployment guide. 

I am suggesting we do it here (but could be persuaded otherwise) that before you create the peer you create the organization msp. It feels disjoint to me to tell them to insert the peer org mspid into the core.yaml even though the msp does not already exist. This is an error prone step that will tank your peer if not done correctly, so I thought it was worth explicitly including it here.

Reviewer: I agree that the org MSP should be created prior to deploying a peer.
However, that should not be WITHIN these instructions, it should be done BEFORE these instructions.
It is in fact described fairly well in the CA deployment guide here: https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#create-the-org-msp-needed-to-add-an-org-to-a-channel
So perhaps we just need to link to that at the BEGINNING of this topic, not WITHIN this topic.

Besides adding the MSPID to the core.yaml peer.localMspId, the org msp directory is not used by the peer at all. So it really doesn't belong in the deploy peer instructions. It is the channel creation/update that uses the org msp directory.

Reviewer 2: OK. That does make sense. Partnered with you other suggestion of creating the 3 sections in the pre-reqs (Certs/storage/cfg), it comes together nicely then. 

I would suggest however that we move these specific instructions into the Fabric CA doc (I am happy to do that).

        - Copy the folders [of your peer organization's MSP](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#create-the-org-msp-needed-to-add-an-org-to-a-channel) to `organizations/peerOrganizations/org1.example.com/msp`. This identifies the organization that owns this peer. **Important:** After you generate the file, remember to edit it and set the path and file name of the root CA certificate for each role. The name of the root CA certificate file is included in the generated MSP under `/msp/cacerts`).

>       - **Peer enrollment certificate and private key**:

Reviewer: Make the bullet name consistent with above, I'd recommend "Peer local MSP (enrollment certificate and private key)".

>         - Next, copy the [local MSP of your node](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#create-the-local-msp-of-a-node) to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp`. This path corresponds to the value of the `peer.mspConfigPath` parameter in the `core.yaml` file. Because of the Fabric concept of ["Node Organization Unit (OU)"](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous), you do do not need to specify an admin of the peer when bootstrapping. Rather, the role of "admin" is conferred onto an identity by setting an OU value of "admin" inside a certificate and enabled by the `config.yaml` file. When Node OUs are enabled, any organization admin will be able to administer the peer.

Reviewer: ```suggestion
        - Next, copy the [local MSP of your node](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#create-the-local-msp-of-a-node) to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp`. This path corresponds to the value of the `peer.mspConfigPath` parameter in the `core.yaml` file. Because of the Fabric concept of ["Node Organization Unit (OU)"](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous), you do not need to specify an admin of the peer when bootstrapping. Rather, the role of "admin" is conferred onto an identity by setting an OU value of "admin" inside a certificate and enabled by the `config.yaml` file. When Node OUs are enabled, any organization admin will be able to administer the peer.
```

        - Note that the local MSP contains the signed certificate (public key) and the private key for the peer. The private key is used by the node to sign transactions, and is therefore not shared and must be secured. For maximum security, a Hardware Security Module (HSM) can be configured to generate and store this private key.

>         - Copy the TLS CA root certificate `ca-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/ca-cert.pem`. The path and name of the certificate corresponds to `peer.tls.clientRootCAs.files` parameter in `core.yaml`. Only required if mutual authentication (`clientAuthRequired`) is set to true. Note that multiple files can be configured, one for each client organization that will communicate with the peer.

Reviewer: When I said that this bullet should be last, I meant last within the **TLS certificate** section, not last within the **Peer enrollment** section.

Author: Whoops!

The resulting folder structure is similar to:

```
├── fabric
  ├── bin
  │   ├── configtxgen
  │   ├── configtxlator
  │   ├── cryptogen
  │   ├── discover
  │   ├── idemixgen
  │   ├── orderer
  │   └── peer
  └── config
    ├── configtx.yaml
    ├── core.yaml
    └── orderer.yaml
```

Along with the relevant binaries, you will receive both the peer binary executable and the peer configuration file, `core.yaml`, that is required to launch a peer on the network. The other files are not required for the peer deployment but will be useful when you attempt to create or edit channels, among other tasks.

**Tip:** Add the location of the peer binary to your `PATH` environment variable so that it can be picked up without fully qualifying the path to the binary executable, for example:

```
export PATH=<path to download location>/bin:$PATH
```

> After you have mastered deploying and running a peer by using these two files, it is likely you will want to use the [peer image](https://hub.docker.com/r/hyperledger/fabric-peer) that Fabric publishes instead, for example in a Kubernetes or Docker deployment. For now though, the purpose of this topic is to teach you how to properly use the peer binary so you can take that knowledge and apply it to the production environment of your choice.

Reviewer: ```suggestion
After you have mastered deploying and running a peer by using the peer binary and core.yaml configuration file, it is likely that you will want to use a peer container in a Kubernetes or Docker deployment. The Hyperledger Fabric project publishes a [peer image](https://hub.docker.com/r/hyperledger/fabric-peer) that can be used for development and test, and various vendors provide supported peer images. For now though, the purpose of this topic is to teach you how to properly use the peer binary so you can take that knowledge and apply it to the production environment of your choice.
```

## Prerequisites

Before you can launch a peer node in a production network, you need to configure it. Complete the following steps:

1. Create the recommended folder structure for the peer certificates that is described in the [Registering and enrolling identities with a CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html) topic to store the generated certificates and MSPs:
    ```
    ├── organizations
      └── peerOrganizations
          └── org1.example.com
              ├── msp
              └── peers
                  └── peer0.org1.example.com
                      ├── msp
                      └── tls
    ```

> 2. While **cryptogen** is a convenient utility that can be used to generate certificates for a test network, it should **never** be used on a production network. The core requirement for certificates for Fabric nodes is that they are Elliptic Curve (EC) certificates. You can use any tool you prefer to issue these certificates (for example, OpenSSL). However, the Fabric CA streamlines the process because it generates the Membership Service Providers (MSPs) for you. Regardless of which certificate provider you choose, you should have already used it to generate the peer enrollment, TLS certificate, private key, and the MSPs that Fabric must consume. Refer to the [CA deployment](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/cadeploy.html) and [Registering and enrolling identities with a CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html) topics for instructions on how to generate these certificates. You need to generate the following sets of certificates:

Reviewer 2: you should have already used it to generate the peer enrollment certificate, TLS certificate, private keys, and the MSPs that Fabric must consume.

    - **Peer TLS certificates**

>     - **Peer organization admin certificates (organization MSP)**

Reviewer: Note that this bullet makes it sound like we'll have an admin cert (or multiple certs). That's not the case when using admin NodeOUs. The NodeOU configuration indicates that ANY cert issued by the org CA with role of 'admin' should be considered an admin.

Author: @pamandrejko I think Dave has a point here and made some edits here pointing to the registering and enrolling CA doc to make some edits here.

Reviewer 2: Yeah. I was using "certificates" to mean certificate + private key. But I am fine changing it to certificate (singular) if that makes more sense here.

    - **Peer enrollment certificates (peer local MSP)**  

> 3. You will either need to use the Fabric CA client to generate the certificates (which are placed into the folder structure Fabric needs to consume) or will need to copy the generated certificates to their designated folders. Whichever method you choose, must users are ultimately likely to script this process so it can be repeatable as needed. A list of the certificates and their locations is provided here for your convenience.

Reviewer 2: You will either need to use the Fabric CA client to generate the certificates directly into the recommended folder structure or you will need to copy the generated certificates to their recommended folders after they are generated.

      - **Peer TLS certificate and private key**:

>         - Copy the TLS CA TLS certificate (`tls-cert.pem`) to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/tls-cert.pem`. The path and name of the certificate corresponds to `tls.rootcert.file` parameter in `core.yaml`.

Reviewer 2: Remove parentheses

>         - Copy the TLS CA TLS certificate `tls-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/tls-cert.pem`. The path and name of the certificate corresponds to `tls.rootcert.file` parameter in `core.yaml`.

Reviewer: peer.tls.rootcert.file default value is "tls/ca.crt". Are you recommending it be overridden?
It is the "TLS CA root certificate", not "TLS CA TLS certificate".
Therefore, copy tls ca.crt file to this location (or whatever file name you want to use).
Note that organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls is incorrect here... that location is only used in "test network". The deployment guide is not for test network.

Author: @pamandrejko Could you weigh in here? Sounds like you transposed this section (which refers to tls.rootcert.file) with the one below it (which talks about the tls root cert).

Reviewer 2: @denyeart  I reviewed this and what I have here is intentional, so correct me if I am wrong. I was using the same naming convention we used in the CA deployment guide - see the diagram at the end of this [topic](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/cadeploy.html#enroll-bootstrap-user-with-tls-ca) which exactly matches the default names of the certs that are generated. 

This is my experience/understanding:

1. `ca-cert.pem`- When you initially start up  a CA server, the file `ca-cert.pem` is generated. This is the self-signed root CA cert. And if this is a TLS CA, it would be the "TLS CA root cert". This file corresponds to the `tls.clientRootCAs.files` parameter.

2. `tls-cert.pem` - When you enable TLS communications on the CA, the file `tls-cert.pem` is generated. (Also why I called it the TLS CA TLS cert which is a horrible name I know).  This file corresponds to the `tls.rootcert.file` parameter. So it does seem like we have a mismatch with your default value - tls/ca.crt. 

I replicated these steps this morning on a new CA server, just to double-check the file naming to make sure I did not transpose anything.  Feel free to correct me.

Reviewer 2: I checked with Saad on this one. I stand corrected:

1) The file that needs to go in `tls.rootcert.file` is the cert that contains the public key that is associated with the signing (private) key, which in a default CA scenario is going to be ca-cert.pem from my example above.

2) `tls.clientRootCAs.files` is only required when mutual TLS is configured. So for tls.clientRootCAs.files it should be the certificate that will be used to verify the client certificate, which file goes there all depends on who is issuing the client’s TLS certificate. Since we are not using mutual TLS in this tutorial, it does not need to be included here.

>         - Copy the TLS CA root certificate `ca-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/ca-cert.pem`. The path and name of the certificate corresponds to `tls.clientRootCAs.files` parameter in `core.yaml`.

Reviewer: peer.tls.clientRootCAs.files is only used for validating client certificates. It will be a list of the client root CAs certs, if clientAuthRequired is true (mutual auth).

>         - Copy the TLS CA root certificate `ca-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/ca-cert.pem`. The path and name of the certificate corresponds to `peer.tls.clientRootCAs.files` parameter in `core.yaml`. Only required if mutual authentication (`clientAuthRequired`) is set to true.

Reviewer: Make this the last bullet, since it relates to client auth only. Also, I thought this would be tls-cert.pem (from the client org TLS CA), not ca-cert.pem.

```suggestion
                - Copy the TLS CA root certificate `ca-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/ca-cert.pem`. The path and name of the certificate corresponds to `peer.tls.clientRootCAs.files` parameter in `core.yaml`. Only required if mutual authentication (`clientAuthRequired`) is set to true. Note that multiple files can be configured, one for each client organization that will communicate with the peer.
```

>         - After you have generated the peer TLS certificate, rename the generated private key in the `keystore` folder to `peer0-key.pem` so that it can more easily be recognized later on as being a private key.

Reviewer: ```suggestion
        - After you have generated the peer TLS certificate, the certificate will have been generated in the `signcerts` directory, and the private key will have been generated in the `keystore` directory. Rename the generated private key in the `keystore` folder to `peer0-key.pem` so that it can more easily be recognized later on as being a private key.
```

>         - Copy the peer TLS certificate and private key to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls`. The path and name of these certificates correspond to the `tls.cert.file` and `tls.key.file` parameters in the `core.yaml`.

Reviewer: peer.tls.key.file default value is "tls/server.key"
peer.tls.cert.file default value is "tls/server.crt".
Again, don't use the test network example.

>         - Copy the peer TLS certificate and private key to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls`. The path and name of these certificates correspond to the `peer.tls.cert.file` and `peer.tls.key.file` parameters in the `core.yaml`.

Reviewer: ```suggestion
        - Copy the peer TLS certificate and private key to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls`. The path and name of the certificate and private key files correspond to the `peer.tls.cert.file` and `peer.tls.key.file` parameters in the `core.yaml`.
```

>       - **Peer organization admin MSP**:

Reviewer: Note that this bullet and the sub-bullets makes it sound like we'll have an admin identity with cert. That's not the case when using admin NodeOUs. The NodeOU configuration indicates that ANY cert issued by the org CA with role of 'admin' should be considered an admin.

Author: See above comment.

>         - After you have registered the peer organization admin user and enrolled the identity, copy the generated MSP folder to `peerOrganizations/org1.example.com/msp`. This is the peer organization MSP definition and is used by every peer node in the organization. This path becomes the value of the `mspConfigPath` parameter in the `core.yaml` file.

Reviewer 2: copy the generated MSP folder to `organizations/peerOrganizations/org1.example.com/msp`

>         - After you have registered the peer organization admin user and enrolled the identity, copy the generated MSP folder to `organizations/peerOrganizations/org1.example.com/msp`. This is the peer organization MSP definition and is used by every peer node in the organization. This path becomes the value of the `mspConfigPath` parameter in the `core.yaml` file.

Reviewer: No, the mspConfigPath is the peer's local msp.
And not sure what you mean by "path becomes the value of the mspConfigPath.

Author: Moved this down to the right place.

>         - After you have registered the peer organization admin user and enrolled the identity, copy the generated MSP folder to `organizations/peerOrganizations/org1.example.com/msp`. This is the peer organization MSP definition and is used by every peer node in the organization.

Reviewer: organizations/peerOrganizations/org1.example.com/msp is for the channel org level msp.
What is being configured here is the peer's local MSP - organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp.
Same with the next bullet... the config.yaml that specifies the admin node ou goes in the peer's local msp directory. 
For this reason, I would make the admin config as the last bullet. That is, you need to define the peer identity (enrollment cert) in the local msp as per the next bullet, and then verify the admin config for that identity.

Author: Hopefully the edits I made address this as well.

>         - Fabric introduced the concept of "Node Organization Unit (OU)" support that allows a role to be conferred onto an identity by setting an OU value inside a certificate. [Node OU](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#specifying-nodeous) support is enabled by creating a `config.yaml` file inside the `/msp` folder in the `organizations/peerOrganizations/org1.example.com` path in which a role is tied to the root of trust for the issued certificate. **Important:** After you generate the file, remember to edit it and set the path and file name of the root CA certificate for each role. The name of the root CA certificate file is included in the generated MSP under `/msp/cacerts`).

Reviewer 2: Let's change the link to https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous

      - **Peer enrollment certificate and private key**:

>         - After you have registered the peer user for this node and enrolled its identity with the peer organization CA, copy the generated `msp` folder and replace the **local MSP** with it under `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/msp`.

Reviewer: Again, don't use the test network location.
The local msp goes in the peer.mspConfigPath location, by default that is /msp.

The resulting folder structure is similar to:

```
├── fabric
  ├── bin
  │   ├── configtxgen
  │   ├── configtxlator
  │   ├── cryptogen
  │   ├── discover
  │   ├── idemixgen
  │   ├── orderer
  │   └── peer
  └── config
    ├── configtx.yaml
    ├── core.yaml
    └── orderer.yaml
```

Along with the relevant binaries, you will receive both the peer binary executable and the peer configuration file, `core.yaml`, that is required to launch a peer on the network. The other files are not required for the peer deployment but will be useful when you attempt to create or edit channels, among other tasks.

**Tip:** Add the location of the peer binary to your `PATH` environment variable so that it can be picked up without fully qualifying the path to the binary executable, for example:

```
export PATH=<path to download location>/bin:$PATH
```

> After you have mastered deploying and running a peer by using these two files, it is likely you will want to use the [peer image](https://hub.docker.com/r/hyperledger/fabric-peer) that Fabric publishes instead, for example in a Kubernetes or Docker deployment. For now though, the purpose of this topic is to teach you how to properly use the peer binary.

Reviewer 2: Suggestion for positioning this content better:

After you have mastered deploying and running a peer by using these two files, it is likely you will want to use the [peer image](https://hub.docker.com/r/hyperledger/fabric-peer) that Fabric publishes instead, for example in a Kubernetes or Docker deployment. For now though, the purpose of this topic is to teach you how to properly use the peer binary so you can take that knowledge and apply it to the production environment of your choice.

## Prerequisites

Before you can launch a peer node in a production network, you need to make sure you've created and organized the necessary certificates, decided on storage, and configured `core.yaml`.

### Certificates

Note: while **cryptogen** is a convenient utility that can be used to generate certificates for a test network, it should **never** be used on a production network. The core requirement for certificates for Fabric nodes is that they are Elliptic Curve (EC) certificates. You can use any tool you prefer to issue these certificates (for example, OpenSSL). However, the Fabric CA streamlines the process because it generates the Membership Service Providers (MSPs) for you.

Before you can deploy the peer, create the recommended folder structure for the peer certificates that is described in the [Registering and enrolling identities with a CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html) topic to store the generated certificates and MSPs.

> This folder structure isn't mandatory, but these instructions presume you have created it:

Reviewer: Not sure why, but the folder structure looks wonky when I view it at https://hyperledger-fabric--1803.org.readthedocs.build/en/1803/deploypeer/peerdeploy.html

    ```
    ├── organizations
      └── peerOrganizations
          └── org1.example.com
              ├── msp
              └── peers
                  └── peer0.org1.example.com
                      ├── msp
                      └── tls
    ```

You should have already used your certificate authority of choice to generate the peer enrollment certificate, TLS certificate, private keys, and the MSPs that Fabric must consume. Refer to the [CA deployment](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/cadeploy.html) and [Registering and enrolling identities with a CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html) topics for instructions on how to create a Fabric CA and how to generate these certificates. You need to generate the following sets of certificates:
    - **Peer TLS CA certificates**

>     - **Peer local MSP (enrollment certificate and private key of the peer)**

Reviewer: These bullets also look wonky at https://hyperledger-fabric--1803.org.readthedocs.build/en/1803/deploypeer/peerdeploy.html

- Copy the TLS certificate that contains the public key that is associated with the signing (private) key certificate, which by default is called `ca-cert.pem`, to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/tls-cert.pem`. The path and name of the certificate corresponds to the `peer.tls.rootcert.file` parameter in `core.yaml`.
- After you have generated the peer TLS certificate, the certificate will have been generated in the `signcerts` directory, and the private key will have been generated in the `keystore` directory. Rename the generated private key in the `keystore` folder to `peer0-key.pem` so that it can more easily be recognized later on as being a private key.
- Copy the peer TLS certificate and private key to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls`. The path and name of the certificate and private key files correspond to the `peer.tls.cert.file` and `peer.tls.key.file` parameters in the `core.yaml`.

> 

Reviewer: Somehow the 4th bullet got lost again, it should go here. And I've updated the bullet text a bit.

- If using mutual authentication (`clientAuthRequired` set to true), you need to indicate to the peer which TLS CA root certificates to use to authorize clients. Copy the organization's TLS CA root certificate `ca-cert.pem` to `organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/ca-cert.pem` so that the organization's clients will be authorized. The path and name of the certificate corresponds to `peer.tls.clientRootCAs.files` parameter in `core.yaml`. Note that multiple files can be configured, one for each client organization that will communicate with the peer (for example if other organizations will use this peer for endorsements).
